stdlib/private/OSLog/OSLogMessage.swift

```diff
@@ -0,0 +1,285 @@
+    preamble = 0
+    argumentCount = 0
+  }
+
+  public mutating func appendLiteral(_ literal: String) {
+    formatString += literal.percentEscapedString
+  }
+
+  /// Define interpolation for expressions of type Int. This definition enables
+  /// passing a formatting option and a privacy qualifier along with the
+  /// interpolated expression as shown below:
+  ///
+  ///         "\(x, as: .hex, is: .private\)"
+  ///
+  /// - Parameters:
+  ///  - number: the interpolated expression of type Int, which is autoclosured.
+  ///  - format: a formatting option available for Int types, defined by the
+  ///    enum `IntFormat`.
+  ///  - privacy: a privacy qualifier which is either private or public.
+  ///    The default is public.
+  public mutating func appendInterpolation(
+    _ number: @autoclosure @escaping () -> Int,
+    as format: IntFormat = .decimal,
+    is privacy: Privacy = .public) {
+    guard argumentCount < maxOSLogArgumentCount else { return }
+
+    addIntHeadersAndFormatSpecifier(
+      format,
+      isPrivate: privacy == .private,
+      byteSize: UInt8(MemoryLayout<Int>.size),
+      isUnsigned: false)
+    argumentSequence.append(number)
+  }
+
+  /// Construct/update format string and headers from the qualifiers (of the
+  /// interpolated expression) passed as parameters.
+  ///
+  /// All arguments to this function must be known at compile time.
+  public mutating func addIntHeadersAndFormatSpecifier(
+    _ format: IntFormat,
+    isPrivate: Bool,
+    byteSize: UInt8,
+    isUnsigned: Bool) {
+    // Always explicitly set either public or private flag bit.
+    let flag: UInt8 = isPrivate ? 1 : 2
+    formatString += getIntegerFormatSpecifier(
+      format,
+      isPrivate: isPrivate,
+      isLong: byteSize > 4,
+      isUnsigned: isUnsigned);
+
+    // Append the argument header and size.
+    argumentSequence.append(flag)
+    argumentSequence.append(byteSize)
+
+    // Update the summary bytes.
+    preamble |= isPrivate ? 1 : 0
+    argumentCount += 1
+  }
+
+  /// Construct an OS log format specifier from the given parameters.
+  /// All arguments to this function must be known at compile time.
+  public func getIntegerFormatSpecifier(
+    _ format: IntFormat,
+    isPrivate: Bool,
+    isLong: Bool,
+    isUnsigned: Bool)
+    -> String {
+    var formatSpecifier: String = isPrivate ? "%{private}" : "%{public}"
+    formatSpecifier += isLong ? "l" : ""
+
+    switch (format) {
+    case .hex:
+      formatSpecifier += "x"
+    case .octal:
+      formatSpecifier += "o"
+    default:
+      formatSpecifier += isUnsigned ? "u" : "d"
+    }
+    return formatSpecifier;
+  }
+}
+
+extension String {
+  /// Replace all percents "%" in the string by "%%" so that the string can be
+  /// interpreted as a C format string.
+  public var percentEscapedString: String {
+    get {
+      return self.reduce(into: "") { (result, char) in
+        result += (char == "%") ? "%%" : String(char)
```

Comment:
It might be faster to first split by `'%'` and then join the result with `"%%"` as a separator. Or it might not...

Reply:
Sounds good to me. (That is definitely more parallel than using `reduce`, if we ever start using parallel collections :-)).